File: src/rithik/card.py
"""The name card printed by `rithik` and `rithik card`.

Every fact here is taken from a public page: the GitHub profile README and the READMEs of the
public repos listed. Numbers are quoted as those pages state them, so the card never claims more
than a reader can check, and nothing links to a private repo.
"""

from __future__ import annotations

import copy
import textwrap

from rithik._term import BOLD, BOLD_CYAN, CYAN, DIM, paint

NAME = "Rithik Krishna T"
TAGLINE = "AI/ML, Android and security. I build complete systems and measure them."

EDUCATION = (
    {
        "degree": "B.Tech CSE (Data Science)",
        "institution": "MGIT Hyderabad",
        "graduation": "May 2027",
    },
    {
        "degree": "BS Data Science & Applications",
        "institution": "IIT Madras",
        "graduation": "May 2027",
    },
)

EXPERIENCE = ({"role": "AI/ML Intern", "organisation": "HCLTech", "period": "Sep 2025 - Jan 2026"},)

LINKS = {
    "github": "https://github.com/Daemon-VI",
    "linkedin": "https://www.linkedin.com/in/rithikkrishnat/",
    "email": "trithikkrishna@gmail.com",
}

PROJECTS = (
    {
        "name": "AegisToolkit",
        "summary": "31-tool Android security toolkit, no root required",
        "url": "https://github.com/Daemon-VI/AegisToolkit",
    },
    {
        "name": "Cascade",
        "summary": "Drag-and-drop Android automation with 154 actions",
        "url": "https://github.com/Daemon-VI/Cascade",
    },
    {
        "name": "optivision-rag",
        "summary": (
            "Visual-token compression for VLM document retrieval: 113.5x less storage per "
            "page, 86.7% of baseline nDCG@5"
        ),
        "url": "https://github.com/Daemon-VI/optivision-rag",
    },
    {
        "name": "carpool",
        "summary": "Shared-commute cost settlement for iOS, Android and web",
        "url": "https://github.com/Daemon-VI/carpool",
    },
    {
        "name": "darkwatch",
        "summary": "Dark web exposure monitor for leak sites and breaches",
        "url": "https://github.com/Daemon-VI/darkwatch",
    },
    {
        "name": "rithik",
        "summary": "This card, plus an offline Indian scam-message checker",
        "url": "https://github.com/Daemon-VI/rithik",
    },
)

HINT_COMMAND = 'rithik scam "<message or link>"'

WIDTH = 78
_INNER = WIDTH - 6  # each row is "|  " + text + "  |"
_LABEL_COL = 13
_PROJECT_COL = 17


def card_data() -> dict:
    # Deep copies, so a caller that edits the result cannot change what the next render prints.
    return {
        "name": NAME,
        "tagline": TAGLINE,
        "education": copy.deepcopy(list(EDUCATION)),
        "experience": copy.deepcopy(list(EXPERIENCE)),
        "links": dict(LINKS),
        "projects": copy.deepcopy(list(PROJECTS)),
    }


def render_card(color: bool = False) -> str:
    # Rows are (plain, styled) pairs: padding has to be measured on the plain text, because
    # escape codes take up characters but no columns.
    rows: list[tuple[str, str]] = [("", "")]
    rows.append((NAME, paint(NAME, BOLD_CYAN, color)))
    rows.extend(_wrapped(TAGLINE))
    rows.append(("", ""))

    degrees = [f"{e['degree']}, {e['institution']}" for e in EDUCATION]
    graduations = {e["graduation"] for e in EDUCATION}
    if len(graduations) == 1:
        degrees.append(f"Both graduating {graduations.pop()}")
    else:
        degrees = [f"{d} ({e['graduation']})" for d, e in zip(degrees, EDUCATION)]
    for i, line in enumerate(degrees):
        rows.extend(_two_col("Education" if i == 0 else "", line, _LABEL_COL, CYAN, color))
    for i, job in enumerate(EXPERIENCE):
        line = f"{job['role']}, {job['organisation']}, {job['period']}"
        rows.extend(_two_col("Experience" if i == 0 else "", line, _LABEL_COL, CYAN, color))
    rows.append(("", ""))

    for label, key in (("GitHub", "github"), ("LinkedIn", "linkedin"), ("Email", "email")):
        rows.extend(_two_col(label, LINKS[key], _LABEL_COL, CYAN, color))
    rows.append(("", ""))

    heading = "Public projects"
    rows.append((heading, paint(heading, BOLD, color)))
    for project in PROJECTS:
        rows.extend(_two_col(project["name"], project["summary"], _PROJECT_COL, BOLD, color))
        rows.extend(_two_col("", project["url"], _PROJECT_COL, BOLD, color))
    rows.append(("", ""))

    border = paint("+" + "-" * (WIDTH - 2) + "+", DIM, color)
    bar = paint("|", DIM, color)
    lines = [border]
    for plain, styled in rows:
        lines.append(f"{bar}  {styled}{' ' * (_INNER - len(plain))}  {bar}")
    lines.append(border)
    lines.append("Try: " + paint(HINT_COMMAND, BOLD, color))
    return "\n".join(lines) + "\n"


def _wrapped(text: str) -> list[tuple[str, str]]:
    return [(piece, piece) for piece in _wrap(text, _INNER)]


def _two_col(
    left: str, right: str, left_width: int, left_code: str, color: bool
) -> list[tuple[str, str]]:
    rows = []
    for i, piece in enumerate(_wrap(right, _INNER - left_width)):
        head = left if i == 0 else ""
        pad = " " * (left_width - len(head))
        rows.append((head + pad + piece, paint(head, left_code, color) + pad + piece))
    return rows


def _wrap(text: str, width: int) -> list[str]:
    # break_on_hyphens=False keeps names like "Drag-and-drop" and URLs in one piece.
    return textwrap.wrap(text, width, break_on_hyphens=False) or [""]

File: js/lib/card.js
// The name card printed by `rithik` and `rithik card`. The port of src/rithik/card.py: keep the
// facts here identical to that file, which says where each one comes from.

import { BOLD, BOLD_CYAN, CYAN, DIM, paint } from "./term.js";
import { cpLength, wrap } from "./pytext.js";

export const NAME = "Rithik Krishna T";
export const TAGLINE = "AI/ML, Android and security. I build complete systems and measure them.";

const EDUCATION = Object.freeze([
  Object.freeze({
    degree: "B.Tech CSE (Data Science)",
    institution: "MGIT Hyderabad",
    graduation: "May 2027",
  }),
  Object.freeze({
    degree: "BS Data Science & Applications",
    institution: "IIT Madras",
    graduation: "May 2027",
  }),
]);

const EXPERIENCE = Object.freeze([
  Object.freeze({ role: "AI/ML Intern", organisation: "HCLTech", period: "Sep 2025 - Jan 2026" }),
]);

const LINKS = Object.freeze({
  github: "https://github.com/Daemon-VI",
  linkedin: "https://www.linkedin.com/in/rithikkrishnat/",
  email: "trithikkrishna@gmail.com",
});

const PROJECTS = Object.freeze([
  Object.freeze({
    name: "AegisToolkit",
    summary: "31-tool Android security toolkit, no root required",
    url: "https://github.com/Daemon-VI/AegisToolkit",
  }),
  Object.freeze({
    name: "Cascade",
    summary: "Drag-and-drop Android automation with 154 actions",
    url: "https://github.com/Daemon-VI/Cascade",
  }),
  Object.freeze({
    name: "optivision-rag",
    summary:
      "Visual-token compression for VLM document retrieval: 113.5x less storage per " +
      "page, 86.7% of baseline nDCG@5",
    url: "https://github.com/Daemon-VI/optivision-rag",
  }),
  Object.freeze({
    name: "carpool",
    summary: "Shared-commute cost settlement for iOS, Android and web",
    url: "https://github.com/Daemon-VI/carpool",
  }),
  Object.freeze({
    name: "darkwatch",
    summary: "Dark web exposure monitor for leak sites and breaches",
    url: "https://github.com/Daemon-VI/darkwatch",
  }),
  Object.freeze({
    name: "rithik",
    summary: "This card, plus an offline Indian scam-message checker",
    url: "https://github.com/Daemon-VI/rithik",
  }),
]);

export const HINT_COMMAND = 'rithik scam "<message or link>"';

const WIDTH = 78;
const INNER = WIDTH - 6; // each row is "|  " + text + "  |"
const LABEL_COL = 13;
const PROJECT_COL = 17;

const copyAll = (items) => items.map((item) => ({ ...item }));

/** The card as plain data, in the key order `rithik card --json` prints. Always a fresh copy. */
export function cardData() {
  return {
    name: NAME,
    tagline: TAGLINE,
    education: copyAll(EDUCATION),
    experience: copyAll(EXPERIENCE),
    links: { ...LINKS },
    projects: copyAll(PROJECTS),
  };
}

export function renderCard(color = false) {
  // Rows are [plain, styled] pairs: padding has to be measured on the plain text, because
  // escape codes take up characters but no columns.
  const rows = [["", ""]];
  rows.push([NAME, paint(NAME, BOLD_CYAN, color)]);
  rows.push(...wrapped(TAGLINE));
  rows.push(["", ""]);

  let degrees = EDUCATION.map((e) => `${e.degree}, ${e.institution}`);
  const graduations = new Set(EDUCATION.map((e) => e.graduation));
  if (graduations.size === 1) {
    degrees.push(`Both graduating ${[...graduations][0]}`);
  } else {
    degrees = degrees.map((d, i) => `${d} (${EDUCATION[i].graduation})`);
  }
  degrees.forEach((line, i) => {
    rows.push(...twoCol(i === 0 ? "Education" : "", line, LABEL_COL, CYAN, color));
  });
  EXPERIENCE.forEach((job, i) => {
    const line = `${job.role}, ${job.organisation}, ${job.period}`;
    rows.push(...twoCol(i === 0 ? "Experience" : "", line, LABEL_COL, CYAN, color));
  });
  rows.push(["", ""]);

  for (const [label, key] of [
    ["GitHub", "github"],
    ["LinkedIn", "linkedin"],
    ["Email", "email"],
  ]) {
    rows.push(...twoCol(label, LINKS[key], LABEL_COL, CYAN, color));
  }
  rows.push(["", ""]);

  const heading = "Public projects";
  rows.push([heading, paint(heading, BOLD, color)]);
  for (const project of PROJECTS) {
    rows.push(...twoCol(project.name, project.summary, PROJECT_COL, BOLD, color));
    rows.push(...twoCol("", project.url, PROJECT_COL, BOLD, color));
  }
  rows.push(["", ""]);

  const border = paint("+" + "-".repeat(WIDTH - 2) + "+", DIM, color);
  const bar = paint("|", DIM, color);
  const lines = [border];
  for (const [plain, styled] of rows) {
    lines.push(`${bar}  ${styled}${spaces(INNER - cpLength(plain))}  ${bar}`);
  }
  lines.push(border);
  lines.push("Try: " + paint(HINT_COMMAND, BOLD, color));
  return lines.join("\n") + "\n";
}

// " " * n in Python is "" for a negative n; String.prototype.repeat would throw.
const spaces = (n) => " ".repeat(Math.max(0, n));

function wrapped(text) {
  return wrapLines(text, INNER).map((piece) => [piece, piece]);
}

function twoCol(left, right, leftWidth, leftCode, color) {
  return wrapLines(right, INNER - leftWidth).map((piece, i) => {
    const head = i === 0 ? left : "";
    const pad = spaces(leftWidth - cpLength(head));
    return [head + pad + piece, paint(head, leftCode, color) + pad + piece];
  });
}

function wrapLines(text, width) {
  // breakOnHyphens: false keeps names like "Drag-and-drop" and URLs in one piece.
  const lines = wrap(text, width, { breakOnHyphens: false });
  return lines.length ? lines : [""];
}
